apps/src/javalab/JavabuilderConnection.js

```diff
@@ -36,50 +35,73 @@ export default class JavabuilderConnection {
     this.setIsTesting = setIsTesting;
     this.executionType = executionType;
     this.miniAppType = miniAppType;
-    this.overrideSources = overrideSources;
   }
 
   // Get the access token to connect to javabuilder and then open the websocket connection.
   // The token prevents access to our javabuilder AWS execution environment by un-verified users.
   connectJavabuilder() {
-    // Don't attempt to connect to Javabuilder if we do not have a project.
+    let requestData = this.getDefaultRequestData();
+    requestData.channelId = this.channelId;
+    const ajaxPayload = {
+      url: '/javabuilder/access_token',
+      type: 'get',
+      data: requestData
+    };
+
+    this.connectJavabuilderHelper(ajaxPayload, /* checkProjectEdited */ true);
+  }
+
+  // Get the access token to connect to javabuilder and then open the websocket connection.
+  // When getting the access token, send override sources to run instead of attempting to find
+  // sources based on a channel id.
+  // The token prevents access to our javabuilder AWS execution environment by un-verified users.
+  connectJavabuilderWithOverrideSources(overrideSources) {
+    let requestData = this.getDefaultRequestData();
+    requestData.overrideSources = overrideSources;
+    const ajaxPayload = {
+      url: '/javabuilder/access_token_with_override_sources',
+      type: 'get',
+      data: requestData
+    };
+
+    // When we have override sources, we do not need to check if the project has been edited,
+    // as the override sources are what we want to run.
+    this.connectJavabuilderHelper(ajaxPayload, /* checkProjectEdited */ false);
+  }
+
+  // Get the access token to connect to javabuilder and then open the websocket connection.
+  // When getting the access token, send override validation code to run instead of any existing validation
+  // code on the level.
+  // The token prevents access to our javabuilder AWS execution environment by un-verified users.
+  connectJavabuilderWithOverrideValidation(overrideValidation) {
+    let requestData = this.getDefaultRequestData();
+    requestData.channelId = this.channelId;
+    requestData.overrideValidation = overrideValidation;
+    const ajaxPayload = {
+      url: '/javabuilder/access_token_with_override_validation',
+      type: 'get',
+      data: requestData
+    };
```

Comment:
nit: instead of creating the full payload each time, `connectJavabuilderHelper` could take in a url and requestData param? it's a small change though so I'm fine with this as well

Reply:
done!